New density operator, bra ket and unitary structures

alexjgreig opened the pull request:
These structures have been added to formalise the elements of a quantum program and to help represent the evolving state in a quantum program.

alexjgreig committed:
Created a new density operator structure to represent the quantum state of the program, additionally added notation for braket notation and for unitary matricies with the respective structures and definitions.

alexjgreig committed:
Finishing structures in unitary, densityOp and quantum files.

alexjgreig commented:
Hi, please review the above and share your thoughts below. Should utilise these structures in implementing the qwhile+ language.

ImHangLi commented on .DS_Store:
Delete this file

ImHangLi commented on Quave/braket.lean:
- I have implemented the conversion between `Kat` and `Bra`, and added comprehensive documentation of the file. You can adapt them all or choose the part that fits. 
- Below is the code:
<details><summary>Details</summary>
<p>

```lean
import Mathlib.Data.Complex.Basic
import Mathlib.Data.Matrix.Basic
import Mathlib.Algebra.BigOperators.Fin
import Mathlib.Data.Matrix.Kronecker
import Mathlib.LinearAlgebra.Matrix.PosDef

/-!
# Bra-Ket Notation for Quantum States

This module implements the bra-ket notation (Dirac notation) for quantum states.
It provides the fundamental structures and operations for quantum state vectors.

## Main definitions

* `Ket` - A normalized column vector representing a quantum state
* `Bra` - The dual vector of a Ket (conjugate transpose)
* `dot` - Inner product between a Bra and a Ket

## Notation

* `|ψ⟩` - Represents a Ket state
* `⟨φ|` - Represents a Bra state
* `⟨φ‖ψ⟩` - Represents the inner product between Bra φ and Ket ψ

## Implementation notes

States are represented as functions from a finite type to complex numbers,
following the convention in `Matrix` of vectors as simple functions.
Both Bra and Ket vectors are required to be normalized (norm = 1).
-/

open Classical
open BigOperators
open ComplexConjugate
open Kronecker
open Complex
open scoped Matrix ComplexOrder

/-! ### Basic definitions -/

noncomputable section

section BraKet
variable (d : Type*) [Fintype d]

/-- A ket as a vector of unit norm. We follow the convention in `Matrix` of vectors as simple functions
 from a Fintype. Kets are distinctly not a vector space in our notion, as they represent only normalized
 states and so cannot (in general) be added or scaled. -/
structure Ket where
  vec : d → ℂ
  normalized' : ∑ x, ‖vec x‖^2 = 1

/-- A bra is definitionally identical to a `Ket`, but are separate to avoid complex conjugation confusion.
 They can be interconverted with the adjoint: `Ket.to_bra` and `Bra.to_ket` -/
structure Bra where
  vec : d → ℂ
  normalized' : ∑ x, ‖vec x‖^2 = 1

end BraKet

/-! ### Conversion between Bra and Ket -/

namespace Ket
variable (d : Type*) [Fintype d]

/-- Convert a Ket to its dual Bra by taking the complex conjugate of each component. -/
def to_bra (ψ : Ket d) : Bra d where
  vec := fun x => conj (ψ.vec x)
  normalized' := by
    rw [← ψ.normalized']
    congr
    funext x
    rw [Complex.norm_eq_abs, Complex.abs_conj]
    rfl
end Ket

namespace Bra
variable (d : Type*) [Fintype d]

/-- Convert a Bra to its dual Ket by taking the complex conjugate of each component. -/
def to_ket (φ : Bra d) : Ket d where
  vec := fun x => conj (φ.vec x)
  normalized' := by
    rw [← φ.normalized']
    congr
    funext x
    rw [Complex.norm_eq_abs, Complex.abs_conj]
    rfl
end Bra

/-! ### Notation and basic operations -/

namespace braket

/-- Notation for Bra vectors using ⟨ψ| -/
scoped notation:max "〈" ψ:90 "∣" => (ψ : Bra _)

/-- Notation for Ket vectors using |ψ⟩ -/
scoped notation:max "∣" ψ:90 "〉" => (ψ : Ket _)

variable {d : Type*} [Fintype d]

/-- Allow using Ket vectors as functions -/
instance instFunLikeKet : FunLike (Ket d) d ℂ where
  coe ψ := ψ.vec
  coe_injective' _ _ h := by rwa [Ket.mk.injEq]

/-- Allow using Bra vectors as functions -/
instance instFunLikeBra : FunLike (Bra d) d ℂ where
  coe ψ := ψ.vec
  coe_injective' _ _ h := by rwa [Bra.mk.injEq]

/-- Inner product between a Bra and a Ket vector -/
def dot (φ : Bra d) (ψ : Ket d) : ℂ := ∑ x, (φ x) * (ψ x)

/-- Notation for inner product using ⟨φ‖ψ⟩ -/
scoped notation "〈" φ:90 "‖" ψ:90 "〉" => dot (φ : Bra _) (ψ : Ket _)

end braket
``` 

</p>
</details> 

- For further changes, consider adding helper functions for creating common states (like |0⟩, |1⟩), and add theorems about normalisation preservation under operations.

ImHangLi commented on Quave/densityop.lean:
- refactor: Convert DensityOp to PartialDensityOp, Because we need this in the qwhile+.
- Below is the code: 
<details><summary>Details</summary>
<p>

```lean
import Mathlib.Data.Matrix.Basic
import Mathlib.Algebra.BigOperators.Fin
import Mathlib.Data.Matrix.Kronecker
import Mathlib.LinearAlgebra.Matrix.PosDef
import Mathlib.LinearAlgebra.Matrix.Trace

/-!
# Partial Density Operators

This module implements partial density operators for quantum states in qWhile+.

## Mathematical Background

A partial density operator ρ has the following properties:
1. Trace(ρ) ≤ 1 (sub-normalization)
2. ρ is positive semidefinite
3. ρ is Hermitian (follows from positive semidefinite)

For a quantum state described by ρ:
* tr(ρ) gives the probability of reaching this state
* ρ/tr(ρ) gives the normalized state when reached
-/

noncomputable section

open Classical BigOperators ComplexConjugate Kronecker Matrix
open scoped Matrix ComplexOrder

/-- A partial density operator represents a quantum state with probability tr(ρ) ≤ 1.
    When normalized by tr(ρ), it gives the actual quantum state. -/
structure PartialDensityOp (d : Type*) [Fintype d] where
  /-- The underlying matrix representation -/
  m : Matrix d d ℂ
  /-- Proof that the matrix is positive semidefinite -/
  pos : m.PosSemidef
  /-- Proof that the trace is at most 1 (sub-normalization) -/
  tr_le_one : m.trace.re ≤ 1

namespace PartialDensityOp

variable {d: Type*} [Fintype d]

/-- Every partial density operator is Hermitian (self-adjoint) -/
theorem Hermitian (ρ : PartialDensityOp d) : ρ.m.IsHermitian :=
  ρ.pos.left


/-- Get the probability of reaching this state -/
def prob (ρ : PartialDensityOp d) : ℝ := ρ.m.trace.re

/-- Get the normalized state ρ/tr(ρ) when probability is non-zero -/
def normalize (ρ : PartialDensityOp d) (h : ρ.prob ≠ 0) : PartialDensityOp d where
  m := (1/ρ.prob) • ρ.m
  pos := sorry  -- TODO: Need to prove (1/prob)•ρ is PSD
  tr_le_one := sorry  -- TODO: Need to prove trace becomes 1

end PartialDensityOp
``` 

</p>
</details> 

This change adapts the density operator implementation to match qWhile+ requirements:

- Renamed DensityOp to PartialDensityOp to reflect its sub-normalized nature
-. Modified trace condition from tr(ρ) = 1 to tr(ρ) ≤ 1
- Added probability interpretation:
   - tr(ρ) represents probability of reaching this state
   - ρ/tr(ρ) gives normalized state when reached
- Added normalize operation to convert partial states to normalized ones

This change is needed to support quantum program semantics where:
- Partial density operators represent quantum states with probabilities
- Measurements and program execution can result in sub-normalized states
- Normalization is required when computing final quantum states

Note: Some proofs are marked as `sorry` and will be completed in follow-up PRs.

ImHangLi commented on Quave/unitary.lean:
- Adapt unitary.lean to the new PartialDensityOp.lean
- Below is the code:
<details><summary>Details</summary>
<p>

```lean
import Quave.PartialDensityOp
import Mathlib.LinearAlgebra.Matrix.Adjugate

/-!
# Unitary Operations on Partial Density Operators

This module implements unitary operations on partial density operators.
Unitary evolution preserves both trace and positive semidefiniteness.
-/

noncomputable section

namespace PartialDensityOp

notation "𝐔[" n "]" => Matrix.unitaryGroup n ℂ

variable {d d₁ d₂ d₃ : Type*}
variable [Fintype d] [Fintype d₁] [Fintype d₂] [Fintype d₃]
variable [DecidableEq d]

/-- Conjugate a state by a unitary matrix (applying the unitary as an evolution).
    This preserves both trace and positive semidefiniteness. -/
def U_conj (ρ : PartialDensityOp d) (U : 𝐔[d]) : PartialDensityOp d where
  m := U * ρ.m * star U
  pos := by
    -- For now, mark as sorry until we can fix the Matrix.PosSemidef issues
    sorry
  tr_le_one := by
    -- Trace is preserved under unitary conjugation
    have h1 : (U * ρ.m * star U).trace = ρ.m.trace := by
      rw [Matrix.trace_mul_cycle]
      simp [Matrix.trace_mul_cycle]
    -- Therefore bound is preserved
    rw [h1]
    exact ρ.tr_le_one

end PartialDensityOp
``` 

</p>
</details>

ImHangLi reviewed (changes requested):
- I've implemented a basic Quantum Measurement structure with:
  - Measurement operators with positive semidefiniteness
  - Completeness relation (∑ₘ Mₘ†Mₘ = I)
  - Probability calculation and post-measurement state update
  - Type* for placeholder type
- Below is the code
<details><summary>Details</summary>
<p>

```lean
import Mathlib.Data.Matrix.Basic
import Mathlib.Data.Matrix.Kronecker
import Mathlib.LinearAlgebra.Matrix.PosDef
import Mathlib.LinearAlgebra.Matrix.Adjugate
import Mathlib.Data.Complex.Basic
import Quave.qtype
import Quave.PartialDensityOp
import Quave.env

/-!
# Quantum Measurements for qWhile+

This module implements quantum measurements as specified in qWhile+.

## Main Definitions

* `adjoint` - Adjoint operation for complex matrices
* `BasicMeasurement` - Core measurement structure with operators and properties
* `prob_outcome` - Probability calculation for measurement outcomes
* `post_measurement` - State update after measurement

## Implementation Notes

Each measurement M consists of:
1. A set of measurement operators {Mₘ} indexed by outcome m
2. Satisfies completeness relation: ∑ₘ Mₘ†Mₘ = I
3. Each operator is positive semidefinite

## TODOs

1. Implement tensor product structure for multi-qubit measurements
2. Add proofs for post-measurement state properties:
   * Positive semidefiniteness
   * Trace less than or equal to 1
3. Add common measurement examples:
   * Computational basis measurement
   * Pauli measurements
4. Integrate with CQState for full program semantics
-/

noncomputable section

open Classical BigOperators Matrix Complex qtype
open scoped Matrix ComplexOrder TensorProduct

/-! ### Helper Functions -/

/-- Adjoint operation for complex matrices.
    For a matrix M, returns M† where (M†)ᵢⱼ = conj(Mⱼᵢ) -/
def adjoint {n m : Type*} [Fintype n] [Fintype m] (M : Matrix n m ℂ) : Matrix m n ℂ :=
  fun i j => conj (M j i)

/-! ### Core Measurement Structure -/

/-- A measurement symbol in qWhile+ with quantum type H and outcome type T.

    For a measurement M:
    * `operators m` gives the measurement operator Mₘ for outcome m
    * `pos m` ensures each operator is positive semidefinite
    * `completeness` ensures ∑ₘ Mₘ†Mₘ = I -/
structure BasicMeasurement (H : Type*) (T : Type*) [Fintype H] [Fintype T] where
  /-- Measurement operators indexed by possible outcomes -/
  operators : T → Matrix H H ℂ
  /-- Each measurement operator is positive semidefinite -/
  pos : ∀ m, Matrix.PosSemidef (operators m)
  /-- Completeness relation: ∑m Mm†Mm = I -/
  completeness : ∑ m, adjoint (operators m) * (operators m) = 1

namespace BasicMeasurement
variable {H : Type*} {T : Type*} [Fintype H] [Fintype T]

/-! ### Measurement Operations -/

/-- Probability of getting outcome m when measuring state ρ.

    Given by p(m) = tr(Mₘ†ρMₘ) -/
def prob_outcome (M : BasicMeasurement H T) (ρ : PartialDensityOp H) (m : T) : ℝ :=
  (adjoint (M.operators m) * ρ.m * (M.operators m)).trace.re

/-- Post-measurement state after outcome m.

    Given by ρₘ = (MₘρMₘ†)/p(m) where p(m) = tr(Mₘ†ρMₘ)

    TODO: Prove
    1. Result is positive semidefinite
    2. Trace is ≤ 1 (tracks probability) -/
def post_measurement (M : BasicMeasurement H T) (ρ : PartialDensityOp H) (m : T)
    (h : prob_outcome M ρ m ≠ 0) : PartialDensityOp H :=
  { m := (1/prob_outcome M ρ m) • (M.operators m * ρ.m * adjoint (M.operators m)),
    pos := sorry,  -- TODO: Prove result is positive semidefinite
    tr_le_one := sorry }  -- TODO: Prove trace ≤ 1

end BasicMeasurement
``` 

</p>
</details> 

 ## TODO
- Add proofs for post-measurement state properties:
   - Positive semidefiniteness
   - Trace ≤ 1
- Implement common measurements (computational basis, Pauli)
- Integrate with CQState for program semantics
- Finish TensorProduct implementation

ImHangLi committed:
Update braket.lean

Implement conversion functions between Bra and Ket. Add comprehensive documentation.

ImHangLi committed:
Create partialdensityop.lean

Change densityop to partialdensityop. Add comprehensive documentation.

ImHangLi committed:
Update unitary.lean

Use partialdensityop.

alexjgreig committed:
Removed DS_Store file and added to gitignore.

alexjgreig committed:
Removed density operator lean file as this is now replaced with the partial density operator.

alexjgreig committed:
Adding measurement lean structure.
